File: packages/forecast/src/rgnr8_forecast/money.py
"""Exact money in integer minor units.

Mirrors the accounting kernel's money model: values are an integer count of a
currency's minor unit (e.g. cents). Python ints are arbitrary precision, so all
arithmetic here is exact and reproducible — never binary floating point.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Iterable

_SCALES: dict[str, int] = {"USD": 2, "EUR": 2, "GBP": 2}


def currency_scale(code: str) -> int:
    """Minor-unit digits for a currency code (2 for USD)."""
    try:
        return _SCALES[code]
    except KeyError as exc:  # pragma: no cover - defensive
        raise ValueError(f"Unknown currency: {code}") from exc


class CurrencyMismatch(ValueError):
    pass


@dataclass(frozen=True, slots=True, order=True)
class Money:
    """An exact monetary amount as integer minor units plus a currency code.

    ``order=True`` compares by (minor_units, currency) — comparisons between
    different currencies are rejected explicitly by the arithmetic helpers, but
    the dataclass ordering is only meaningful within one currency.
    """

    minor_units: int
    currency: str = "USD"

    # --- constructors --------------------------------------------------------
    @staticmethod
    def zero(currency: str = "USD") -> "Money":
        return Money(0, currency)

    @staticmethod
    def from_decimal(value: str | int, currency: str = "USD") -> "Money":
        """Parse an exact decimal string (e.g. ``"1234.56"``) into minor units.

        Rejects more fractional digits than the currency allows rather than
        silently rounding — precision loss must be an explicit caller decision.
        """
        scale = currency_scale(currency)
        s = str(value).strip()
        neg = s.startswith("-")
        if neg:
            s = s[1:]
        if "." in s:
            whole, frac = s.split(".", 1)
        else:
            whole, frac = s, ""
        if not whole.isdigit() or (frac and not frac.isdigit()):
            raise ValueError(f"Not a valid decimal amount: {value!r}")
        if len(frac) > scale:
            raise ValueError(
                f"Amount {value!r} has more precision than {currency} allows (scale {scale})"
            )
        frac = frac.ljust(scale, "0")
        minor = int(whole or "0") * (10**scale) + int(frac or "0")
        return Money(-minor if neg else minor, currency)

    # --- arithmetic ----------------------------------------------------------
    def _same(self, other: "Money") -> None:
        if self.currency != other.currency:
            raise CurrencyMismatch(f"{self.currency} vs {other.currency}")

    def __add__(self, other: "Money") -> "Money":
        self._same(other)
        return Money(self.minor_units + other.minor_units, self.currency)

    def __sub__(self, other: "Money") -> "Money":
        self._same(other)
        return Money(self.minor_units - other.minor_units, self.currency)

    def __neg__(self) -> "Money":
        return Money(-self.minor_units, self.currency)

    def scale_by(self, numerator: int, denominator: int) -> "Money":
        """Multiply by an exact rational ``numerator/denominator`` with banker's
        rounding to the nearest minor unit. Used for haircuts and fractions."""
        if denominator == 0:
            raise ValueError("denominator must be non-zero")
        num = self.minor_units * numerator
        den = denominator
        # Normalise the sign onto the numerator so the denominator is positive.
        # `divmod` then floors toward -inf with 0 <= r < den, so `q` is always the
        # lower integer and rounding up is unconditionally `q += 1` — regardless of
        # the value's sign. (The previous `q -= 1` branch double-rounded negatives.)
        if den < 0:
            num, den = -num, -den
        q, r = divmod(num, den)
        twice = r * 2
        # round half to even
        if twice > den or (twice == den and q % 2 != 0):
            q += 1
        return Money(q, self.currency)

    def abs(self) -> "Money":
        return Money(abs(self.minor_units), self.currency)

    # --- predicates ----------------------------------------------------------
    @property
    def is_zero(self) -> bool:
        return self.minor_units == 0

    @property
    def is_positive(self) -> bool:
        return self.minor_units > 0

    @property
    def is_negative(self) -> bool:
        return self.minor_units < 0

    # --- formatting ----------------------------------------------------------
    def to_decimal_string(self) -> str:
        scale = currency_scale(self.currency)
        neg = self.minor_units < 0
        s = str(abs(self.minor_units)).rjust(scale + 1, "0")
        if scale == 0:
            body = s
        else:
            body = f"{s[:-scale]}.{s[-scale:]}"
        return f"-{body}" if neg else body

    def __str__(self) -> str:
        return f"{self.currency} {self.to_decimal_string()}"


def money_sum(items: Iterable[Money], currency: str = "USD") -> Money:
    total = Money.zero(currency)
    for m in items:
        total = total + m
    return total

File: packages/ledger-kernel/src/money.ts
/**
 * Money — exact monetary values in integer minor units.
 *
 * INVARIANT: money is never represented as a binary floating-point number.
 * All values are stored as a `bigint` count of the currency's minor unit
 * (e.g. cents for USD). Arithmetic is therefore exact and reproducible.
 */

export interface Currency {
  /** ISO 4217-style code, e.g. "USD". */
  readonly code: string;
  /** Number of minor-unit digits, e.g. 2 for USD (cents). */
  readonly scale: number;
}

/** Registry of known currencies, so a stored currency code can be rehydrated. */
const CURRENCY_REGISTRY = new Map<string, Currency>();

export function defineCurrency(code: string, scale: number): Currency {
  if (!/^[A-Z]{3}$/.test(code)) {
    throw new Error(`Invalid currency code: ${code}`);
  }
  if (!Number.isInteger(scale) || scale < 0 || scale > 8) {
    throw new Error(`Invalid currency scale for ${code}: ${scale}`);
  }
  const existing = CURRENCY_REGISTRY.get(code);
  if (existing) {
    if (existing.scale !== scale) {
      throw new Error(`Currency ${code} already defined with scale ${existing.scale}`);
    }
    return existing;
  }
  const currency = Object.freeze({ code, scale });
  CURRENCY_REGISTRY.set(code, currency);
  return currency;
}

/** Resolve a currency by code. Throws if it has not been defined. */
export function getCurrency(code: string): Currency {
  const c = CURRENCY_REGISTRY.get(code);
  if (!c) throw new Error(`Unknown currency code: ${code}. Define it with defineCurrency() first.`);
  return c;
}

export const USD: Currency = defineCurrency("USD", 2);

export class CurrencyMismatchError extends Error {
  override readonly name = "CurrencyMismatchError";
  constructor(a: Currency, b: Currency) {
    super(`Currency mismatch: ${a.code} vs ${b.code}`);
  }
}

const DECIMAL_RE = /^(-)?(\d+)(?:\.(\d+))?$/;

export class Money {
  private constructor(
    readonly minorUnits: bigint,
    readonly currency: Currency,
  ) {
    Object.freeze(this);
  }

  static fromMinorUnits(amount: bigint | number, currency: Currency): Money {
    const v = typeof amount === "number" ? BigInt(assertSafeInteger(amount)) : amount;
    return new Money(v, currency);
  }

  static zero(currency: Currency): Money {
    return new Money(0n, currency);
  }

  /**
   * Parse an exact decimal string (e.g. "1234.56") into minor units.
   * Rejects more fractional digits than the currency scale rather than
   * silently rounding — precision loss must be an explicit caller decision.
   */
  static fromDecimal(value: string, currency: Currency): Money {
    const m = DECIMAL_RE.exec(value.trim());
    if (!m) throw new Error(`Not a valid decimal amount: "${value}"`);
    const sign = m[1] === "-" ? -1n : 1n;
    const whole = m[2] ?? "0";
    const frac = m[3] ?? "";
    if (frac.length > currency.scale) {
      throw new Error(
        `Amount "${value}" has more precision than ${currency.code} allows (scale ${currency.scale})`,
      );
    }
    const paddedFrac = frac.padEnd(currency.scale, "0");
    const combined = BigInt(whole) * 10n ** BigInt(currency.scale) + BigInt(paddedFrac || "0");
    return new Money(sign * combined, currency);
  }

  plus(other: Money): Money {
    this.assertSameCurrency(other);
    return new Money(this.minorUnits + other.minorUnits, this.currency);
  }

  minus(other: Money): Money {
    this.assertSameCurrency(other);
    return new Money(this.minorUnits - other.minorUnits, this.currency);
  }

  negate(): Money {
    return new Money(-this.minorUnits, this.currency);
  }

  abs(): Money {
    return this.minorUnits < 0n ? this.negate() : this;
  }

  /** Multiply by an integer factor (e.g. quantity). Kept integer-only on purpose. */
  timesInteger(factor: bigint | number): Money {
    const f = typeof factor === "number" ? BigInt(assertSafeInteger(factor)) : factor;
    return new Money(this.minorUnits * f, this.currency);
  }

  isZero(): boolean {
    return this.minorUnits === 0n;
  }

  isPositive(): boolean {
    return this.minorUnits > 0n;
  }

  isNegative(): boolean {
    return this.minorUnits < 0n;
  }

  equals(other: Money): boolean {
    return this.currency.code === other.currency.code && this.minorUnits === other.minorUnits;
  }

  /** -1, 0, or 1. Throws on currency mismatch. */
  compare(other: Money): -1 | 0 | 1 {
    this.assertSameCurrency(other);
    if (this.minorUnits < other.minorUnits) return -1;
    if (this.minorUnits > other.minorUnits) return 1;
    return 0;
  }

  lessThan(other: Money): boolean {
    return this.compare(other) < 0;
  }

  greaterThan(other: Money): boolean {
    return this.compare(other) > 0;
  }

  toDecimalString(): string {
    const neg = this.minorUnits < 0n;
    const abs = neg ? -this.minorUnits : this.minorUnits;
    const s = abs.toString().padStart(this.currency.scale + 1, "0");
    const cut = s.length - this.currency.scale;
    const whole = s.slice(0, cut);
    const frac = this.currency.scale > 0 ? "." + s.slice(cut) : "";
    return `${neg ? "-" : ""}${whole}${frac}`;
  }

  toString(): string {
    return `${this.currency.code} ${this.toDecimalString()}`;
  }

  private assertSameCurrency(other: Money): void {
    if (this.currency.code !== other.currency.code) {
      throw new CurrencyMismatchError(this.currency, other.currency);
    }
  }
}

/** Sum a list of Money, all of the same currency. Empty list requires a currency. */
export function sumMoney(items: readonly Money[], currency?: Currency): Money {
  if (items.length === 0) {
    if (!currency) throw new Error("sumMoney: empty list requires an explicit currency");
    return Money.zero(currency);
  }
  return items.reduce((acc, m) => acc.plus(m));
}

function assertSafeInteger(n: number): number {
  if (!Number.isInteger(n)) throw new Error(`Expected an integer, got ${n}`);
  if (!Number.isSafeInteger(n)) throw new Error(`Integer ${n} exceeds safe range; pass a bigint`);
  return n;
}
